src/input_candidates.rs

```diff
@@ -0,0 +1,185 @@
+use std::{collections::BTreeSet, vec::Vec};
+
+use crate::{collections::BTreeMap, DefiniteDescriptor, InputGroup, InputStatus};
+use bdk_chain::{
+    local_chain::LocalChain, BlockId, ConfirmationBlockTime, KeychainIndexed, TxGraph,
+};
+use bitcoin::{absolute, OutPoint, ScriptBuf};
+use core::fmt::{Debug, Display};
+use miniscript::{bitcoin, plan::Assets, Descriptor, DescriptorPublicKey, ForEachKey};
+
+use crate::Input;
+
+/// Input candidates that are not processed (filtered and/or grouped).
+///
+/// Some inputs may be unspendable now (due to unsatisfied time-locks for they are immature
+/// coinbase spends).
+///
+/// TODO: This should live in `bdk_chain` after we move `Input`, `InputGroup`, types to
+/// `bdk_tx_core`.
+#[derive(Debug, Clone)]
+pub struct InputCandidates {
+    inputs: Vec<Input>,
+}
+
+/// Error
+#[derive(Debug)]
+pub enum InputCandidatesError<K> {
+    /// Descriptor is missing for keychain K.
+    MissingDescriptor(K),
+    /// Cannot plan descriptor. Missing assets?
+    CannotPlan(DefiniteDescriptor),
+}
+
+impl<K: Debug> Display for InputCandidatesError<K> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            InputCandidatesError::MissingDescriptor(k) => {
+                write!(f, "missing descriptor for keychain {:?}", k)
+            }
+            InputCandidatesError::CannotPlan(descriptor) => {
+                write!(f, "cannot plan input with descriptor {}", descriptor)
+            }
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl<K: Debug> std::error::Error for InputCandidatesError<K> {}
+
+/// Default group policy.
+pub fn group_by_spk(input: &Input) -> ScriptBuf {
+    input.prev_txout().script_pubkey.clone()
+}
+
+/// No grouping.
+pub fn no_groups(input: &Input) -> OutPoint {
+    input.prev_outpoint()
+}
+
+/// Filter out inputs that cannot be spent now.
+pub fn filter_unspendable_now(
+    tip_height: absolute::Height,
+    tip_time: absolute::Time,
+) -> impl Fn(&InputGroup) -> bool {
+    move |group| group.is_spendable_now(tip_height, tip_time)
+}
+
+impl InputCandidates {
+    /// Construct.
+    ///
+    /// # Error
+    ///
+    /// Requires a descriptor for each corresponding K value.
+    pub fn new<K>(
+        tx_graph: &TxGraph<ConfirmationBlockTime>,
+        chain: &LocalChain,
+        chain_tip: BlockId,
+        outpoints: impl IntoIterator<Item = KeychainIndexed<K, OutPoint>>,
+        descriptors: BTreeMap<K, Descriptor<DescriptorPublicKey>>,
+        allow_malleable: BTreeSet<K>,
+        additional_assets: Assets,
+    ) -> Result<Self, InputCandidatesError<K>>
+    where
+        K: Clone + Ord + Debug,
+    {
+        let mut pks = vec![];
+        for desc in descriptors.values() {
+            desc.for_each_key(|k| {
+                pks.extend(k.clone().into_single_keys());
+                true
+            });
+        }
```

Comment:
As mentioned in [@ValuedMammal's HackMD](https://hackmd.io/@valuedmammal/B1lxzomj0#UX), sometimes the caller may not have all keys available for spending from. What is the right API so that it is not tedious to fill in available assets?